File: canandmessage/canandmessage_translingual/canandmessage_parser/toml_defs.py
import typing
from .serde import Serde, Anything

def default_true() -> bool:
    return True

def default_false() -> bool:
    return False

def default_uint() -> str:
    return "uint"

def default_scale() -> list :
    return [1,1]

#[derive(Deserialize, Debug, Clone)]
class DeviceSpec(Serde):
    name: str
    base: typing.List[str]
    arch: str
    #[serde(default = "default_true")]
    is_public: bool = default_true
    dev_type: int
    dev_class: int
    msg: typing.Dict[str, 'DeviceMessageSpec'] = dict
    settings: typing.Dict[str, 'DeviceSettingSpec'] = dict
    #[serde(default = "BTreeMap::new")]
    types: typing.Dict[str, 'TypeSpec'] = dict
    #[serde(default = "BTreeMap::new")]
    enums: typing.Dict[str, 'EnumSpec'] = dict
    #[serde(default = "BTreeMap::new")]
    setting_commands: typing.Dict[str, 'SettingCommandSpec'] = dict

    vendordep: typing.Optional['VendordepSpec']

#[derive(Deserialize, Debug, Clone)]
class VendordepSpec(Serde):
    java_package: str
    cpp_namespace: str

#[derive(Deserialize, Debug, Clone)]
class DeviceMessageSpec(Serde):
    id: int
    min_length: typing.Optional[int]
    max_length: typing.Optional[int]
    length: typing.Optional[int]
    frame_period_setting: typing.Optional[str]
    source: str
    #[serde(default = "default_true")]
    is_public: bool = default_true
    #[serde(default = "default_true")]
    vendordep: bool = default_true
    comment: str
    signals: typing.List['MessageSignalSpec']

#[derive(Deserialize, Debug, Clone)]
class MessageSignalSpec(Serde):
    name: str
    comment: str
    dtype: str
    #[serde(default = "bool::default")]
    optional: bool = default_false
    default_value: Anything

    #[serde(default = "bool::default")]
    mux: bool = default_false
    muxed_by: typing.Optional[str]
    muxed_match: Anything

#[derive(Deserialize, Debug, Clone)]
class DeviceSettingSpec(Serde):
    id: int
    comment: str
    dtype: str
    default_value: Anything


    #[serde(default = "default_true")]
    is_public: bool = default_true
    #[serde(default = "default_true")]
    vendordep: bool = default_true
    #[serde(default = "default_true")]
    vdep_setting: bool = default_true
    #[serde(default = "default_true")]
    readable: bool = default_true
    #[serde(default = "default_true")]
    writable: bool = default_true
    #[serde(default = "default_true")]
    reset_on_default: bool = default_true
    #[serde(default = "Vec::default")]
    special_flags: typing.List[str] = list


#[derive(Deserialize, Debug, Clone)]
class TypeSpec(Serde):
    btype: str
    comment: str = str
    unit: str = str

    #[serde(default = "String::default")]
    utype: str = str
    bits: int
    min: Anything
    max: Anything
    #[serde(default = "default_true")]
    allow_nan_inf: bool = default_true
    default_value: Anything
    #[serde(default = "default_scale")]
    factor: list = default_scale
    offset: Anything
    #[serde(default = "Vec::default")]
    signals: typing.List[MessageSignalSpec] = list
    #[serde(default = "Vec::default")]
    bit_flags: typing.List['BitsetFlagSpec'] = list

#[derive(Deserialize, Debug, Clone)]
class SettingCommandSpec(Serde):
    id: int
    #[serde(default = "default_true")]
    vendordep: bool = default_true
    comment: str

#[derive(Deserialize, Debug, Clone)]
class BitsetFlagSpec(Serde):
    name: str
    comment: str

#[derive(Deserialize, Debug, Clone)]
class EnumSpec(Serde):
    #[serde(default = "default_uint")]
    comment: str = str
    btype: str = default_uint
    bits: int
    #[serde(default = "default_true")]
    is_public: bool = default_true
    default_value: str
    values: typing.Dict[str, 'EnumEntrySpec']

#[derive(Deserialize, Debug, Clone)]
class EnumEntrySpec(Serde):
    id: int
    comment: str

File: canandmessage/canandmessage_parser/src/toml_defs.rs
use serde::Deserialize;
use std::collections::BTreeMap;
use toml::Value;

fn default_true() -> bool {
    true
}
fn default_uint() -> String {
    String::from("uint")
}

fn default_scale() -> [i64; 2] {
    [1, 1]
}
#[derive(Deserialize, Debug, Clone)]
pub struct DeviceSpec {
    pub name: String,
    pub base: Vec<String>,
    pub arch: String,
    #[serde(default = "default_true")]
    pub is_public: bool,
    pub dev_type: u8,
    pub dev_class: u8,
    pub msg: BTreeMap<String, DeviceMessageSpec>,
    #[serde(default = "BTreeMap::new")]
    pub settings: BTreeMap<String, DeviceSettingSpec>,
    #[serde(default = "BTreeMap::new")]
    pub types: BTreeMap<String, TypeSpec>,
    #[serde(default = "BTreeMap::new")]
    pub enums: BTreeMap<String, EnumSpec>,
    #[serde(default = "BTreeMap::new")]
    pub setting_commands: BTreeMap<String, SettingCommandSpec>,

    pub vendordep: Option<VendordepSpec>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct VendordepSpec {
    pub java_package: String,
    pub cpp_namespace: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct DeviceMessageSpec {
    pub id: u8,
    pub min_length: Option<u8>,
    pub max_length: Option<u8>,
    pub length: Option<u8>,
    pub source: String,
    #[serde(default = "default_true")]
    pub is_public: bool,
    #[serde(default = "default_true")]
    pub vendordep: bool,
    pub comment: String,
    pub signals: Vec<MessageSignalSpec>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct MessageSignalSpec {
    pub name: String,
    pub comment: String,
    pub dtype: String,
    #[serde(default = "bool::default")]
    pub optional: bool,
    pub default_value: Option<Value>,

    #[serde(default = "bool::default")]
    pub mux: bool,
    pub muxed_by: Option<String>,
    pub muxed_match: Option<Value>, // TODO: this isn't correct

    #[serde(default = "default_true")]
    pub alchemist: bool,
}

#[derive(Deserialize, Debug, Clone)]
pub struct DeviceSettingSpec {
    pub id: u8,
    pub comment: String,
    pub dtype: String,
    pub default_value: Option<Value>,

    #[serde(default = "default_true")]
    pub is_public: bool,
    #[serde(default = "default_true")]
    pub vendordep: bool,
    #[serde(default = "default_true")]
    pub vdep_setting: bool,
    #[serde(default = "default_true")]
    pub readable: bool,
    #[serde(default = "default_true")]
    pub writable: bool,
    #[serde(default = "default_true")]
    pub reset_on_default: bool,
    #[serde(default = "Vec::default")]
    pub special_flags: Vec<String>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct TypeSpec {
    pub btype: String,
    #[serde(default = "String::default")]
    pub utype: String,

    #[serde(skip, default = "String::default")]
    pub origin_lname: String,

    pub bits: u8,
    pub min: Option<Value>,
    pub max: Option<Value>,
    #[serde(default = "default_true")]
    pub allow_nan_inf: bool,
    pub default_value: Option<Value>,
    #[serde(default = "default_scale")]
    pub factor: [i64; 2],
    pub offset: Option<Value>,
    #[serde(default = "Vec::default")]
    pub signals: Vec<MessageSignalSpec>,
    #[serde(default = "Vec::default")]
    pub bit_flags: Vec<BitsetFlagSpec>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct SettingCommandSpec {
    pub id: u8,
    #[serde(default = "default_true")]
    pub vendordep: bool,
    pub comment: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct BitsetFlagSpec {
    pub name: String,
    pub comment: String,
}

#[derive(Deserialize, Debug, Clone)]
pub struct EnumSpec {
    #[serde(default = "default_uint")]
    pub btype: String,
    pub bits: u8,
    #[serde(default = "default_true")]
    pub is_public: bool,
    #[serde(skip, default = "String::default")]
    pub origin_lname: String,
    pub default_value: String,
    pub values: BTreeMap<String, EnumEntrySpec>,
}

#[derive(Deserialize, Debug, Clone)]
pub struct EnumEntrySpec {
    pub id: u32,
    pub comment: String,
}
